packages/neuron-ui/src/components/SignAndVerify/index.tsx

```diff
@@ -181,11 +183,16 @@ const SignAndVerify = () => {
 
     getAddressesByWalletID(id).then(res => {
       if (isSuccessResponse(res)) {
-        setWallet({ id, addresses: res.result })
+        setWallet(v => ({ ...v, id, addresses: res.result }))
       } else {
         showErrorMessage(t('messages.error'), typeof res.message === 'string' ? res.message : res.message.content!)
       }
     })
+    getCurrentWallet().then(res => {
```

Comment:
With these codes, we can remove the `getAddressesByWalletID` call above, since the addresses are limited to the current wallet and they are returned by `getCurrentWallet`, even more, the `current wallet` has been returned by `useGlobalState()` [line 164]

Reply:
should i delete `const currentWallet = useMemo(() => wallets.find(w => w.id === wallet?.id), [wallet, wallets])` by instead of `state.wallet`

Reply:
I think it's fine to delete this hook
I've confirmed 2 points:
1. the `id` in URL is bound to the current wallet on opening the window and won't change https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-wallet/src/controllers/app/menu.ts#L304;
2. the `wallet` in sign-and-verify window is set by `id` in the URL;
3. the sign-and-verify window will close when the current wallet changes https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L168

so there's no need to update the `wallet` in sign-and-verify window again.

Reply:
I've found something new

There are 2 `current wallet` in this component, `wallet` in https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L158 and `currentWallet` in https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L166
They may be different since `currentWallet` is expected to have `device` property https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L339

Reply:
`state.wallet` is same as  `currentWallet`,they don't have `addresses`.So i should use `state.wallet` instead of `currentWallet` and retain `wallet` in https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L158 . Or i return `addresses` by `get-current-wallet`, then i can only use `state.wallet` in this component.

Reply:
> `state.wallet` is same as `currentWallet`,they don't have `addresses`.So i should use `state.wallet` instead of `currentWallet` and retain `wallet` in [https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L158](https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx?rgh-link-date=2022-03-22T06%3A33%3A56Z#L158) . Or i return `addresses` by `get-current-wallet`, then i can only use `state.wallet` in this component.

I'm not sure which one is specified as `state.wallet` that don't have `addresses`, the one returned by `useGlobalState` has `addresses` but has no `device`  property while the `settings.wallets` returned by `useGlobalState` has the `device` property. I think that's the reason `currentWallet` is filtered from `settings.wallets`

Reply:
Typeof `state.wallet` include `addresses`, but actual it has no `addresses`. I think why `State.Wallet` has addresses property because used at [this](https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L158). The `state.wallet` returned by `get-current-wallet` or `load-init-data` all don't has `addresses`.Because they all returned by https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-wallet/src/services/wallets.ts#L338. The `Wallet` class has no addresses property.So i think i should add `addresses`.

`state.wallet` actual has `device` property when create Hardware wallet at https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-wallet/src/controllers/wallets.ts#L258

Reply:
My point may be misunderstood.

`addresses` could be fetched from `wallet` returned by `useGlobalState()`(https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L162-L164) so there's no need to use `getAddressesByWalletID` to fetch it again.

But the `wallet` returned by `useGlobalState` doesn't have the `device` property. However `settings.wallets` returned by `useGlobalState()` have, so the `device` property should be fetched by `settings.wallets.filter(w => w.id === wallet.id)?.device`

Reply:
`The wallet returned by useGlobalState() doesn't have the device property.` I don't have hardware wallet, i think it has device property by `useGlobalState`.The device property has saved when create wallet. https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-wallet/src/controllers/wallets.ts#L258.

Reply:
> Typeof `state.wallet` include `addresses`, but actual it has no `addresses`

I missed the first point.

Actually `state.wallet` has `addresses`, I can log in the console

Reply:
> `The wallet returned by useGlobalState() doesn't have the device property.` I don't have hardware wallet, i think it has device property by `useGlobalState`.The device property has saved when create wallet.
> 
> https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-wallet/src/controllers/wallets.ts#L258
> 
> .

Indeed, the `wallet` returned by `useGlobalState` has the `device` property and will be instantiated as `HardwareWallet` if `device` exists.

Reply:
> 

sorry, I missed the code.`wallet.addresses` assign by https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-ui/src/states/stateProvider/actionCreators/app.ts#L31 or https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-ui/src/states/init/wallet.ts#L16

Reply:
> > `The wallet returned by useGlobalState() doesn't have the device property.` I don't have hardware wallet, i think it has device property by `useGlobalState`.The device property has saved when create wallet.
> > https://github.com/nervosnetwork/neuron/blob/33bb20311fdce1e99803104a48492b8863eadaba/packages/neuron-wallet/src/controllers/wallets.ts#L258
> > 
> > .
> 
> Indeed, the `wallet` returned by `useGlobalState` has the `device` property and will be instantiated as `HardwareWallet` if `device` exists.

yes, it's really has `device` property.So i can delete https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-ui/src/components/SignAndVerify/index.tsx#L166 and use `useGlobalState().wallet` to replace
![image](https://user-images.githubusercontent.com/12881040/159615426-8c91898d-50e6-4bc3-a0c5-e44015171b37.png)